### ISSUE TITLE
Unit test coverage needs to be increased and maintained

### ISSUE THREAD

**Issue by emmagarland** [umbraco/Umbraco-CMS#9904]
### Issue/suggestion summary

In my talk about the member work in the recent UmbLondon I mentioned that some of the members area wasn't covered by unit tests, meaning I had lower confidence in the changes I was making. 

As we have known for a while, it is clear that unit test coverage needs to be increased. And then maintained so it doesn't drop again.  Along with the cleanup discussed in #9396.

After a quick discussion with @lars-erik on the chat, I was thinking about an approach similar to the way we've been listing other .NET Core tasks to do. 

## Suggested approach

- We could run Jetbrains DotCover and work out the areas that have lowest test coverage, and create smaller tasks from those to add test coverage
- We can add lists of check-boxes TODO items (seems to work fine in GitHub) and people can comment when they're picking an area up
- The new tests can also follow builder pattern discussed by @AndyButland for the construction of model classes used in tests (see #7898)

## Umbraco version

I think this would be best to start in the `netcore\dev` project. 

This relates to #9396, #7898 and possibly other issues already raised. 

### Expected outcome

- Gradually ticking off increasing unit test coverage via pull requests
- Increasing test coverage to a reasonable amount, e.g. a confirmed percentage we increase and then maintain on the solution moving forwards

Comments and suggestions welcome!

**Comment by nzdev** [comment]
It would be good for a similar effort on V8, or to write tests on v8 then port the tests to .net core as it's still supported.

**Comment by emmagarland** [comment]
> It would be good for a similar effort on V8, or to write tests on v8 then port the tests to .net core as it's still supported.

I agree, we would definitley benefit from the unit tests in both. It is just a case of where to start, as a lot of the tests have been reorganised. Probably one to discuss with @bergmania too.

**Comment by bergmania** [comment]
I agree it would be nice to have for both v8 and netcore, but I'm unsure what code can be reused due to the big changes in our tests, pipeline setup etc.

**Comment by emmagarland** [comment]
@bergmania true - I originally thought starting the project to increase coverage would be on netcore since we've done so much in there already to organise the tests. Any other ideas welcome!

**Comment by lars-erik** [comment]
My two cents is that the unit tests have long been sub par and both new and old things have been hard to test. A lot of stubbing is done to avoid integration, often long before the actual integration takes place. To me it's a definite symptom of architecture blocking proper GUTs. (Good unit testst)

As far as I've seen there's been quite a push against more testable architecture in V9. @nzdev' and @rustybox' contributions comes to mind. I don't think even half of that is backportable, so I think the thoughts of improving V8 coverage / tests is kind of futile. I've ranted enough about that in other fora, and most recently this in hopes of better testability in V9. #10106

#7898 may go long way towards being able to running the tests both integrated and not only by swapping configuration. (Environment config even)

I realize I'm slightly off topic addressing the entire suite of tests, including the content cache. The membership stuff might be slightly "better" already and backportable. My main point is that .NET Core, better architecture and all that will ultimately give us way better coverage of use cases with Membership as well. It will sadly not be that nice a ride for V8 because the public interfaces are fixed.

**Comment by emmagarland** [comment]
@lars-erik thanks, so I guess it sounds like the most logical action next is to progress with increasing and maintaining the unit test coverage on v9? 

Would it be a good start if I looked at creating tasks for the **lowest areas of coverage** in v9, and starting smaller tasks from there, based on the test framework that has been built up?

cc @bergmania @AndyButland

**Comment by rustybox** [comment]
Coverage is an odd metric like loc counts (lines of code).

Important areas need thorough testing, unimportant areas less so.

I’m pretty out of the loop these days, have the namespace cleanups taken place to make #9396 less painful?

**Comment by lars-erik** [comment]
I wholeheartedly agree to @rustybox' comment about using it as a metric.  
I'm a huge fan of "use-case" based tests.  
Looking at coverage to find untested "use-cases" is awesome.  
Looking at coverage to make tests called:  
public void Test6() { // Tests the case when parameter x.Y.Z is null  
is more wasteful. :P

So yes, look at coverage, but find untested _use-cases_. :)  
I think that will automatically increase coverage to "acceptable levels" somewhere between 50% and 90%.

**Comment by emmagarland** [comment]
Sounds like a great place to start - and I totally agree re not just adding use cases just for the sake of it @rustybox . 

I will have to check but I believe work has been done to update the namespaces and containing folders for the unit tests in v9.

OK, so like you say @lars-erik the actual next step is to find some key **untested use cases** and increase the coverage from there. 

It would be good to have some sort of coverage metric to start with so we can see progress and also help us maintain it and avoid it dropping over time.
